File: solutions/javascript/rna-transcription/1/rna-transcription.js
//
// This is only a SKELETON file for the 'RNA Transcription' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const toRna = (dna) => {
  // Handle empty string
  if (!dna) {
    return '';
  }

  // Create a mapping object for DNA to RNA conversion
  const transcription = {
    'G': 'C',
    'C': 'G',
    'T': 'A',
    'A': 'U'
  };

  // Convert each nucleotide and join back into a string
  let rna = '';
  
  for (let i = 0; i < dna.length; i++) {
    const nucleotide = dna[i].toUpperCase();
    
    // Check if nucleotide is valid
    if (!transcription.hasOwnProperty(nucleotide)) {
      throw new Error('Invalid input DNA.');
    }
    
    rna += transcription[nucleotide];
  }
  
  return rna;
};

File: solutions/python/rna-transcription/1/rna_transcription.py
def to_rna(dna_strand: str) -> str:
    """
    Transcribe a DNA strand to its RNA complement.
    
    Args:
        dna_strand: A string containing DNA nucleotides (A, C, G, T)
    
    Returns:
        A string containing RNA nucleotides (A, C, G, U)
    
    Raises:
        ValueError: If the input contains invalid DNA nucleotides
    """
    # Mapping dictionary for DNA to RNA transcription
    transcription_map = {
        'G': 'C',
        'C': 'G', 
        'T': 'A',
        'A': 'U'
    }
    
    # Validate input and transcribe
    result = []
    for nucleotide in dna_strand:
        if nucleotide not in transcription_map:
            raise ValueError(f"Invalid DNA nucleotide: {nucleotide}")
        result.append(transcription_map[nucleotide])
    
    return ''.join(result)